New dedicated PM page, with a new "Behavior" section (#12)
multipath-tcp/mptcp.dev: pm.md

---
layout: home
title: Path-Manager
nav_order: 3
nav_titles: true
titles_max_depth: 2
---

The Path Manager is in charge of *subflows*, from creation to deletion, and also
address announcements. Typically, it is the client side that initiates subflows,
and the server side that announces additional addresses via the `ADD_ADDR` and
`REMOVE_ADDR` options.

```mermaid
graph LR;
    C_1(<div style="display: inline-block; min-width: 35px"><font size="7">fa:fa-mobile</font></div>)
    S_1((<div style="display: inline-block; min-width: 60px"><font size="7">fa:fa-cloud</font></div>))

    C_1 -. "Potential subflow" -.- S_1
    C_1 <== "Initial subflow" ==> S_1
    C_1 ~~~|"Subflows creation"| C_1

>     S_1 ~~~|"Addresses announcement"| S_1

Reviewer: ```suggestion
    S_1 ~~~|"Address announcement"| S_1
```

#### Example

- Servers can announce extra IP addresses:
```sh
ip mptcp endpoint add 10.2.2.2 dev eth0 signal
```

- Clients can create additional subflows from a cellular interface, and flag
  this subflow as "backup", to be used to carry data only if the main path is
  unavailable:
```sh
ip mptcp endpoint add 100.64.1.134 dev usb0 subflow backup
```

#### Limits

It is also important to make sure the limits are high enough:

```sh
ip mptcp limits set [ subflows NR ] [ add_addr_accepted NR ]
```

- `subflows` is the limit of additional created and accepted subflows (paths),
  for both the client and server sides (default is 2).
- `add_addr_accepted` is the limit of accepted `ADD_ADDR` -- IP address
  notification from the other peer -- that will result in the creation of
  subflows, typically only for the client side (default is 0).

The limits are per MPTCP connection, and netns-aware.

{: .note}
It is possible to reach the limits with fewer established subflows than
expected, e.g. when new subflow requests cannot reach the other peer. In case of
problem, please increase the limits, use `ss -Mai` to check the counters, and

> set the adapted routing or firewall rules to avoid using certain paths between
> specific IP addresses, e.g. in a lab setup with dedicated links, don't use
> default but specific routes.

Reviewer: Sync with #11 suggestion:

```suggestion
modify the adapted routing or firewall rules to avoid using certain paths between
specific IP addresses. For example, in a lab setup with dedicated links use specific routes 
rather than letting the kernel select the default route.
```

### Behavior

By default, the Path-Manager tries not to create too many subflows using the
encoded [endpoints](#endpoints), and following the imposed [limits](#limits).

{: .note}
Technically, the Path-Manager can only attach new subflows, and announce
addresses on MPTCP connections that are considered as *fully established*:
typically when the first data has been sent.

#### Announcing new addresses

All endpoints flagged as [`signal`](#endpoints) will be announced via an
`ADD_ADDR` notification.

> This will be done, once at a time: once the MPTCP connection is *fully

Reviewer: ```suggestion
This will be done, one at a time: once the MPTCP connection is *fully
```

established*, and each time a new subflow is established. This behaviour might
change in the future, if someone implements the ticket
[#334](https://github.com/multipath-tcp/mptcp_net-next/issues/334).

#### Creating new subflows

> There are two cases that involves the creation of new subflows, if allowed by
> the `subflows` [limits](#limits):

Reviewer: ```suggestion
There are two cases that involve the creation of new subflows, if allowed by
the `subflows` [limit](#limits):
```

- Endpoints flagged as [`subflow`](#endpoints) will be used to create new
  subflows, one at a time, to the address of the server used in the initial

>   subflow. Each endpoint should only be used once per subflow, except if the

Reviewer: ```suggestion
  subflow. Each endpoint should only be used once, except if the
```
I think this is the intended meaning? Or maybe "once per MPTCP connection"?

  endpoint has the [`fullmesh`](#endpoints) flag.

- Upon the reception of an [`ADD_ADDR`](#announcing-new-addresses), and if the
  `add_addr_accepted` [limits](#limits) has not been reached yet, a new subflow
  will be created using the local address the routing configuration will pick,
  except if there are endpoints with the [`fullmesh`](#endpoints) flag. In this
  case, each endpoints with the [`fullmesh`](#endpoints) flag will be used to
  create a new subflow to the announced address.

Note that when subflows are closed before the end of a connection -- e.g. due to
an error on the network, or if the other peer closed subflows -- the
Path-Manager will not try to re-establish them. This behaviour might change in
the future, if someone implements the ticket
[#440](https://github.com/multipath-tcp/mptcp_net-next/issues/440).

#### Accepting new subflows

If the request is valid, the peer will accept the creation of new subflows, as

> long as the `subflows` [limits](#limits) has not been reached yet.

Reviewer: ```suggestion
long as the `subflows` [limit](#limits) has not been reached yet.
```
